hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/slowlog/SlowLogEventHandler.java

```diff
@@ -0,0 +1,205 @@
+          }
+        }
+      }
+    }
+    final String userName = rpcCall.getRequestUserName().orElse(StringUtils.EMPTY);
+    final String methodDescriptorName =
+      methodDescriptor != null ? methodDescriptor.getName() : StringUtils.EMPTY;
+    SlowLogPayload slowLogPayload = SlowLogPayload.newBuilder()
+      .setCallDetails(methodDescriptorName + "(" + param.getClass().getName() + ")")
+      .setClientAddress(clientAddress)
+      .setMethodName(methodDescriptorName)
+      .setMultiGets(numGets)
+      .setMultiMutations(numMutations)
+      .setMultiServiceCalls(numServiceCalls)
+      .setParam(slowLogParams != null ? slowLogParams.getParams() : StringUtils.EMPTY)
+      .setProcessingTime(processingTime)
+      .setQueueTime(qTime)
+      .setRegionName(slowLogParams != null ? slowLogParams.getRegionName() : StringUtils.EMPTY)
+      .setResponseSize(responseSize)
+      .setServerClass(className)
+      .setStartTime(startTime)
+      .setUserName(userName)
+      .build();
+    queue.add(slowLogPayload);
+  }
+
+  /**
+   * Cleans up slow log payloads
+   *
+   * @return true if slow log payloads are cleaned up, false otherwise
+   */
+  boolean clearSlowLogs() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Received request to clean up online slowlog buffer..");
+    }
+    queue.clear();
+    return queue.isEmpty();
+  }
+
+  /**
+   * Retrieve list of slow log payloads
+   *
+   * @param request slow log request parameters
+   * @return list of slow log payloads
+   */
+  List<SlowLogPayload> getSlowLogPayloads(final AdminProtos.SlowLogResponseRequest request) {
+    List<SlowLogPayload> slowLogPayloadList =
+      Arrays.stream(queue.toArray(new SlowLogPayload[0])).collect(Collectors.toList());
+    Collections.reverse(slowLogPayloadList);
+    if (isFilterProvided(request)) {
+      slowLogPayloadList = filterSlowLogs(request, slowLogPayloadList);
+    }
+    int limit = request.getLimit() >= slowLogPayloadList.size() ? slowLogPayloadList.size()
+      : request.getLimit();
+    return slowLogPayloadList.subList(0, limit);
+  }
+
+  private boolean isFilterProvided(AdminProtos.SlowLogResponseRequest request) {
+    if (request.hasUserName() && StringUtils.isNotEmpty(request.getUserName())) {
+      return true;
+    }
+    if (request.hasTableName() && StringUtils.isNotEmpty(request.getTableName())) {
+      return true;
+    }
+    if (request.hasClientAddress() && StringUtils.isNotEmpty(request.getClientAddress())) {
+      return true;
+    }
+    return request.hasRegionName() && StringUtils.isNotEmpty(request.getRegionName());
+  }
+
+  private List<SlowLogPayload> filterSlowLogs(AdminProtos.SlowLogResponseRequest request,
+      List<SlowLogPayload> slowLogPayloadList) {
+    List<SlowLogPayload> filteredSlowLogPayloads = new ArrayList<>();
+    for (SlowLogPayload slowLogPayload : slowLogPayloadList) {
+      if (request.hasRegionName() && StringUtils.isNotEmpty(request.getRegionName())) {
```

Comment:
hasFoo() should not be removed for `optional` fields in protobuf. Regardless of empty check(which we want anyways), hasFoo() is a must, nothing we can do to avoid it.

Reply:
https://developers.google.com/protocol-buffers/docs/proto#optional 

Optional field's default value is empty string. has checks can be skipped for this case..